### PR TITLE
deprecate qk layer scaling and fp32 softmax args

### PR DESCRIPTION
**Pull request by ksivaman**

### CODE REVIEW

File: transformer_engine/pytorch/softmax.py [hunk 1]
```diff
@@ -205,8 +205,8 @@ def __init__(
         self,
         attn_mask_type: str,
         mask_func: Callable,
-        softmax_in_fp32: bool,
         scale: float,
+        softmax_in_fp32: bool = True,
```
Comment (ptrendx) [inline transformer_engine/pytorch/softmax.py:209]: Why did you change the order?

File: transformer_engine/pytorch/transformer.py [hunk 2]
```diff
@@ -82,9 +84,6 @@ def __init__(
         norm_factor: float,
         attention_dropout: float = 0.0,
         attention_dropout_ctx: Optional[Callable] = nullcontext,
-        layer_number: Optional[int] = None,
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:85]: I'm not sure about removal of this layer number. Even if that is just a numerical hack, we should probably do it (just automatically) when needed, and then this would be needed.

File: transformer_engine/pytorch/softmax.py (hunk 1 above)
Comment (ksivaman) [reply transformer_engine/pytorch/softmax.py:209]: To make `softmax_in_fp32` optional as opposed to completely removing it

File: transformer_engine/pytorch/transformer.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:85]: It is a completely unused trick that was needed at a much earlier time to train large models in fp16. No other attention implementation (including FA) has this. MLM team said that it is no longer needed too.

File: transformer_engine/pytorch/softmax.py [hunk 3]
```diff
@@ -260,7 +260,9 @@ def forward_fused_softmax(
     ) -> torch.Tensor:
         """Fused masked softmax kernel"""
         b, np, sq, sk = inp.size()
-        scale = self.scale if self.scale is not None else 1.0
+        scale = 1.0
+        if self.scale is not None and self.input_in_fp16:
```
Comment (ptrendx) [inline transformer_engine/pytorch/softmax.py:264]: I don't like this - I would not want function named "scaled_softmax" to apply the scale I provided only in certain cases, this is surprising and will lead to bugs. It should be the responsibility of the caller of that softmax to either provide or not provide the scale depending on the input type.

File: transformer_engine/pytorch/transformer.py [hunk 4]
```diff
@@ -95,12 +95,12 @@ def __init__(
...
         self.norm_factor = norm_factor
         self.attention_dropout_ctx = attention_dropout_ctx
+        self.layer_number = layer_number
 
         self.scale_mask_softmax = FusedScaleMaskSoftmax(
             attn_mask_type,
             attention_mask_func,
-            attention_softmax_in_fp32,
-            layer_number if apply_query_key_layer_scaling else None,
+            layer_number,
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:103]: Maybe you need to move that scale to the forward call of the softmax so that you know what type its input is going to be.

Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:103]: And then pass `None` if it is not fp16.

**Review by ptrendx — APPROVED** [review]
Lgtm
